keps/sig-storage/20200120-skip-permission-change.md

```diff
@@ -0,0 +1,147 @@
+---
+title: Skip Volume Ownership Change
+authors:
+  - "@gnuified"
+owning-sig: sig-storage
+participating-sigs:
+  - sig-auth
+reviewers:
+  - "@msau42"
+  - "@liggit"
+  - "@tallclair"
+approvers:
+  - "@saad-ali"
+editor: TBD
+creation-date: 2020-01-20
+last-updated: 2020-01-20
+status: implementable
+see-also:
+replaces:
+superseded-by:
+
+---
+
+# Skip Volume Ownership Change
+
+## Table of Contents
+
+<!-- toc -->
+- [Summary](#summary)
+- [Motivation](#motivation)
+  - [Goals](#goals)
+  - [Non-Goals](#non-goals)
+- [Proposal](#proposal)
+  - [Implementation Details/Notes/Constraints [optional]](#implementation-detailsnotesconstraints-optional)
+  - [Risks and Mitigations](#risks-and-mitigations)
+- [Graduation Criteria](#graduation-criteria)
+  - [Test Plan](#test-plan)
+  - [Monitoring](#monitoring)
+- [Implementation History](#implementation-history)
+- [Drawbacks [optional]](#drawbacks-optional)
+- [Alternatives [optional]](#alternatives-optional)
+- [Infrastructure Needed [optional]](#infrastructure-needed-optional)
+<!-- /toc -->
+
+## Summary
+
+Currently before a volume is bind-mounted inside a container the permissions on
+that volume are changed recursively to the provided fsGroup value.  This change
+in ownership can take an excessively long time to complete, especially for very
+large volumes (>=1TB) as well as a few other reasons detailed in [Motivation].
+To solve this issue we will add a new field called `pod.Spec.SecurityContext.VolumePermissionChangePolicy` and
+allow the user to specify whether they want the ownership change to occur.
+
+## Motivation
+
+When a volume is mounted on the node, we recursively change permissions of volume
+before bind mounting the volume inside container. The reason of doing this is to ensure
+that volumes are readable/writable by provided fsGroup.
+
+But this presents following problems:
+ - An application(many popular databases) which is sensitive to permission bits changing
+   underneath may refuse to start whenever volume being used inside pod gets mounted on
+   different node.
```

Comment:
Why does changing nodes make it change permissions?

Reply:
Lets say postgres expects the permission on files managed/created by it to be - `640`  but when the pod moves to another node, it will be `660` and we have seen cases where pod will not start correctly on another node.

Reply:
Ok I see, there's a distinction between ownership and permission. Should we clearly distinguish the difference, or just call both of them "permission" changes?

Reply:
I have already used distinct language in bunch of places, but I will clarify more.  esp - if we are going to allow just ownership change but not permission bits change.

Reply:
in SetVolumeOwnership function, besides it uses Chown to change the gid, it also uses chmod to change the permission mode to 0660 ( or 0440 if readonly). I think the reason behind it is to allow user to access the volume with group permission. If it is still 0640, then a user still cannot access the volume with gid is set because uid is different.

Reply:
They could still run with fsGroup if userid remains the same. which is not too hard to do.

Reply:
userid remains the same? Do you mean the file uid and runAsUser is the same id? then why they need to set fsGroup?